### DISCUSSION TITLE
Equations of States Methods Missing

### DISCUSSION THREAD

**Discussion by JonathanEdelmanS** [idaholab/moose#28920] [Q&A Modules: General]
I am trying to simulate the flow of CO2 through a pipe, using CO2's properties instead of Helium in the example. Are these fluid properties being worked on to be made more complete?

I would like to use this class, and before implementing the missing method myself I would like to know if anybody else is working on it, or otherwise if I do implement the method, how can I contribute that to MOOSE?

When trying to run the code, I get the following error:

`The fluid properties class 'CO2FluidProperties' has not implemented the method below. If your application requires this method, you must either implement it or use a different fluid properties class.

virtual Real SinglePhaseFluidProperties::e_from_p_rho(Real, Real) const`

Below is the .i file, which is modules/thermal_hydraulics/tutorials/single_phase_flow/01_flow_channel.i
where I changed the fluid property from an IdealGasFluidProperties to a CO2FluidProperties:

```T_in = 300. # K
m_dot_in = 1e-2 # kg/s
press = 10e5 # Pa

[GlobalParams]
  initial_p = ${press}
  initial_vel = 0.0001
  initial_T = ${T_in}
  gravity_vector = '0 0 0'

  rdg_slope_reconstruction = minmod
  scaling_factor_1phase = '1 1e-2 1e-4'
  closures = thm_closures
  fp = he
[]

[FluidProperties]
  [he]
    type = CO2FluidProperties
  []
[]

[Closures]
  [thm_closures]
    type = Closures1PhaseTHM
  []
[]

[Components]
  [inlet]
    type = InletMassFlowRateTemperature1Phase
    input = 'core_chan:in'
    m_dot = ${m_dot_in}
    T = ${T_in}
  []

  [core_chan]
    type = FlowChannel1Phase
    position = '0 0 0'
    orientation = '0 0 1'
    length = 1
    n_elems = 25
    A = 7.2548e-3
    D_h = 7.0636e-2
  []

  [outlet]
    type = Outlet1Phase
    input = 'core_chan:out'
    p = ${press}
  []
[]

[Postprocessors]
  [core_p_in]
    type = SideAverageValue
    boundary = core_chan:in
    variable = p
  []

  [core_p_out]
    type = SideAverageValue
    boundary = core_chan:out
    variable = p
  []

  [core_delta_p]
    type = ParsedPostprocessor
    pp_names = 'core_p_in core_p_out'
    expression = 'core_p_in - core_p_out'
  []
[]

[Preconditioning]
  [pc]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  line_search = basic

  start_time = 0
  end_time = 1000
  dt = 10

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-8
  nl_max_its = 25

[]

[Outputs]
  exodus = true

  [console]
    type = Console
    max_rows = 1
    outlier_variable_norms = false
  []
  print_linear_residuals = false
[]```

**Comment by GiudGiud** [comment]
Hello

These routines are implemented on a per need basis. 

For co2 the better option here is to download the external fluid property submodule for carbon dioxide. See the website page for external FPS for more details

**Reply by JonathanEdelmanS** [reply]
Thank you for your response!

We want to simulate that the CO2 is injected at the temperature of liquid storage and is pumped downhole so it reaches supercritical state at a depth of hundreds of meters. We want to detect phase changes or the formation of hydrates,  but we don't need to accurately model phase change, only detect possible phase changes from a default gas/supercritical state.

Would these external fluid properties accurately model supercritical CO2?

**Reply by GiudGiud** [reply]
the external CO2 fluid property module contains properties for both liquid and vapor CO2. I think it's your best bet
https://github.com/idaholab/carbon_dioxide/tree/devel

I'm not sure for supercritical. Is the HEM model good for that? What do you need that's different from a vapor model that's extended up to the condensation limit?
